src/zeit/retresco/content.py

```diff
@@ -0,0 +1,199 @@
+import UserDict
+import grokcore.component as grok
+import lxml.objectify
+import os.path
+import zeit.cms.interfaces
+import zeit.connector.interfaces
+import zeit.retresco.interfaces
+import zope.component
+import zope.schema.interfaces
+
+
+class Content(object):
+
+    zope.interface.implements(zeit.retresco.interfaces.ITMSContent)
+
+    uniqueId = None
+    __name__ = None
+
+    def __init__(self, data):
+        self._tms_payload = data.get('payload', {})
+        if 'url' in data:
+            self.uniqueId = zeit.cms.interfaces.ID_NAMESPACE + data['url'][1:]
+            self.__name__ = os.path.basename(self.uniqueId)
+        self._build_xml_body()
+        self._build_xml_head()
+
+    def _build_xml_body(self):
+        E = lxml.objectify.E
+        self.xml = E.content()
+        for key, value in self._tms_payload.get('xml', {}).items():
+            self.xml.append(getattr(E, key)(value))
+        for container in ['body', 'teaser']:
+            if container not in self._tms_payload:
+                continue
+            container = getattr(E, container)()
+            self.xml.append(container)
+            for key, value in self._tms_payload[container.tag].items():
+                container.append(getattr(E, key)(value))
+
+    def _build_xml_head(self):
+        """head items must be handled explicitly, because of their structure
+        and type conversion, which we need so they are properly queryable.
+
+        We duplicate knowledge about the XML format here for performance
+        reasons. If we wanted to use vivi APIs we first would have to resolve
+        any referenced ICMSContent; instead we can create XML nodes directly.
+        """
+        if 'head' not in self._tms_payload:
+            return
+        data = self._tms_payload['head']
+
+        E = lxml.objectify.E
+        head = E.head()
+        self.xml.append(head)
+
+        for id in data.get('authors', ()):
+            # See zeit.content.author.reference.XMLReference
+            head.append(E.author(href=id))
+
+        image = data.get('teaser_image')
+        if image:
+            # See zeit.content.image.imagegroup.XMLReference
+            image = E.image(**{'base-id': image})
+            fill_color = data.get('teaser_image_fill_color')
+            if fill_color:
+                image.set('fill_color', fill_color)
+            head.append(image)
+
+        if 'covers' in data:
+            # See zeit.content.volume.volume.Volume.set_cover
+            covers = E.covers()
+            self.xml.append(covers)
+            for cover in data['covers']:
+                covers.append(E.cover(**cover))
+
+
+@grok.adapter(dict)
+@grok.implementer(zeit.retresco.interfaces.ITMSContent)
+def from_tms_result(context):
```

Comment:
is this about getting the TMS properties into the WebDAV output? if so, this might be worth a comment or two. if not even more so :)

Reply:
Ne, das ist die Public API um aus einem TMS-Ergebnis ein Content-Objekt zu machen. Das dokumentiere ich noch am Interface.